paasta_tools/kubernetes_tools.py

```diff
@@ -1378,29 +1379,64 @@ def list_deployments(
 @async_timeout()
 async def get_tail_lines_for_kubernetes_pod(
     kube_client: KubeClient, pod: V1Pod, num_tail_lines: int
-) -> MutableMapping[str, List[str]]:
-    tail_lines_dict: MutableMapping[str, Any] = {
-        "stdout": [],
-        "stderr": [],
-        "error_message": "",
-    }
-    for container in pod.spec.containers:
-        if container.name != HACHECK_POD_NAME:
-            try:
-                tail_lines_dict["stdout"].append(
-                    kube_client.core.read_namespaced_pod_log(
-                        name=pod.metadata.name,
-                        namespace="paasta",
-                        container=container.name,
-                        tail_lines=num_tail_lines,
-                    )
-                )
-            except ApiException as e:
-                tail_lines_dict[
-                    "error_message"
-                ] = f"couldn't read stdout/stderr because {e.getResponseBody()}"
+) -> Sequence[MutableMapping[str, Any]]:
+    container_tail_lines: Sequence[MutableMapping[str, Any]] = []
 
-    return tail_lines_dict
+    for container in pod.status.container_statuses:
+        if container.name == HACHECK_POD_NAME:
+            continue
+
+        if container.state.waiting:
+            error = container.state.waiting.message
+        elif container.state.terminated:
+            error = container.state.terminated.message
+        error = error or ""  # if no errors, message will None. convert to ''
+
+        lines = {
+            "container": container.name,
+            "stdout": [],
+            "stderr": [],
+            "error_message": error,
+        }
+        try:
+            log = kube_client.core.read_namespaced_pod_log(
+                name=pod.metadata.name,
+                namespace=pod.metadata.namespace,
+                container=container.name,
+                tail_lines=num_tail_lines,
+            )
+            lines["stdout"].extend(log.split("\n"))
+        except ApiException as e:
+            # there is a potential race condition in which a pod's containers
+            # have not failed, but have when we get the container's logs. in this
+            # case, use the error from the exception, though it is less accurate.
```

Comment:
What's if we move the check for container.state above here? In this way, we won't have race condition here, right?

Reply:
It won't because the state check is on the current in-memory pod object, which can fall behind the actual pod object in Kubernetes. Getting the pod/container logs with `read_namespaced_pod_log` does not update the pod object we have. It's the same race condition as PaaSTA having a pod object for a deleted pod, meaning PaaSTA thinks it exists but k8s has already moved on.

In order to get an update, we need to make a call to the k8s API to retrieve, which I don't think is worth it here. Users can just call `paasta status` again to get the update.

Reply:
Ah, got it.